### DISCUSSION TITLE
Mesh Generation Workshop 2 Planning

### DISCUSSION THREAD

**Discussion by SorooshMani-NOAA** [oceanmodeling/Management#22] [General]
We'd like to hold another meeting/workshop to share ideas about mesh generation and mesh generation tools for coastal modeling purposes. We opened this discussion to gather all the ideas for the workshop planning.

**Comment by SorooshMani-NOAA** [comment]
@WPringle @tomsail @pmav99 @brey had a brief meeting on 3/19 to come up with some initial ideas. As opposed to last workshop/meeting (https://github.com/oceanmodeling/seamesh/tree/main/presentations) where we didn't have a tangible outcome, this time we'd like to be able to come up with ideas for everyone involved to start developing and contributing.

For planning this meeting 3 main questions need to be answered:
- Who's the audience?
- How long the meeting is (half a day, full day, multiple days)?
- What does the meeting focus on?

### Who
I think it makes more sense if this time we only include **willing modelers**, i.e. users of the meshing tool that would also contribute to its development either in terms of code development or ideas and feature list. We already have a good idea of what meshing engines are out there and what their capabilities are, so there's no need to involve mesh engine (triangulation) developers.

### Topics
The following items need to be considered/discussed in the workshop:
- Existing tools' features and missing pieces
- Solver support (e.g. SCHISM, WW3, MPAS, etc.)
- List and pros and cons of each engine
  + Also support for wrapping multiple engines
- Wishlist and features such as
  + Global vs regional mesh
  + Euclidean vs Ellipsoid meshing
  + Projection/CRS handling
  + Feature extraction
  + Quality metrics and mesh post processing
- Development
  + Should we start development from scratch or continue on an existing tool
  + Modular system
     * Should DEM processing be separate from mesh package?
  + Avoiding one package does all idea (e.g. [pyPoseidon](https://github.com/ec-jrc/pyPoseidon))
  + Internal mesh representation
  + IO can be handed off (plugins, etc.) to other packages (e.g. [`meshio`](https://github.com/nschloe/meshio))
- Usability
  + User inputs
  + Use cases

### Format
Depending on the amount of time we dedicate to the workshop, we might be able to tackle discussing both a list of existing tools and workflows (at a practical hands-on level) as well as developing wishlist for where to go next.

Ideas:
- Start by presenting existing tools
- The presenters also prepare a **list of features** they use, and **wishlist** of what they would like to see (bullet points)
- After all presentations we gather all lists and start cataloging features, what feature exists where and whether this is a good candidate to move to the target repo
- We can prepare a code base for the workshop and then discuss evolving it as a part of workshop.

**Comment by tomsail** [comment]
Hi all, 

I'll initiate the discussion with our use case at the JRC: 
we generate **global meshes** for global **surge** models. The models we use are: 
- [TELEMAC](https://gitlab.pam-retd.fr/otm/telemac-mascaret)
- [SCHISM](https://github.com/schism-dev/schism)

We have encapsulated 3 meshers inside [pyPoseidon](https://github.com/ec-jrc/pyPoseidon): 
* [jigsaw](https://github.com/dengwirda/jigsaw)
* [gmsh](https://gitlab.onelab.info/gmsh/gmsh)
* [oceanmesh](https://github.com/CHLNDDEV/oceanmesh/tree/master/oceanmesh)

Here are the pro/cons for each mesher: 

| Tool | Pros | Cons |
|------|------|------|
| Jigsaw | <ul><li> Very efficient in performance</li><li> Quality of the mesh is really good</li></ul> | <ul><li> License not compatible with GPL</li></ul> |
| Gmsh | <ul><li>CLI & Python wrapper</li><li>Possibility to define local scale function on the boundary independently</li><li>Support for multiple scaling</li><li>Large active community with extensive documentation</li><li> GPL license</li></ul> | <ul><li>Takes time to run for complex and big coastlines</li></ul> |
| OceanMesh | <ul><li>Very modular (all sizing functions can be switch on/off)</li><ul><li>Distance</li><li>Feature</li><li>Bathy gradient</li><li>Wavelength</li><li>Thalweg</li></ul><li>Good quality meshes</li><li>Should work for different projections</li></ul> | <ul><li> Small community on Python</li></ul> |

### Metrics / Wish-list
To analyse and correct the mesh, the following metrics would be nice to have to prepare the mesh (depending on the solver you use)

- Equilateralness
- Max neighbor node number
- Max size or area
- CFL condition

TELEMAC built-in [check-mesh.py](https://gitlab.pam-retd.fr/otm/telemac-mascaret/-/blob/main/scripts/python3/check_mesh.py?ref_type=heads) routine has also the following checks that could be interesting for other models: 
- Circle_ratio
- Check_angle
- Over-constrained nodes
- Orthogonality

pyPoseidon has also a few check/correction implemented: 
- Check the mesh against the coastlines
- A CCW check (triangle orientation)
- Triangles crossing the IDL (international date line)

One pager presented in the last OceanModeling meeting with useful links
[OM_management_meeting_20240606.pdf](https://github.com/user-attachments/files/16172955/OM_management_meeting_20240606.1.pdf)

Other software with interesting features: 
 * [BlueKenue](https://nrc.canada.ca/en/research-development/products-services/software-applications/blue-kenuetm-software-tool-hydraulic-modellers) for generating, editing, visualising Selafin (TELEMAC format) files. 
 * [MeshTools](https://github.com/Artelia/Mesh_tools) (by @daoum , @nicogodet and @ccoulet): QGIS plugin for visualizing mesh quality (Equilateralness / Max  number of neighbors, Max size or area). This one could be interesting for you also @AliS-Noaa 

--- 
## Global model implementations
* pyposeidon - oceanmesh > reproducible versions available at: https://github.com/seareport/seareport_models
* pyposeidon - gmsh > example available at https://github.com/ec-jrc/pyPoseidon/blob/master/Tutorial/GLOBAL.ipynb
* jigsaw (for WW3 models) https://github.com/NOAA-EMC/WW3-tools/tree/develop/unst_msh_gen
* seamsh example: https://git.immc.ucl.ac.be/jlambrechts/seamsh/-/blob/master/examples/6-stereographics.py?ref_type=heads
